File: codes/array/54.spiral-matrix/spiral-matrix.go
package main

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func spiralOrder(matrix [][]int) []int {
	sizeY := len(matrix)
	sizeX := len(matrix[0])
	ans := make([]int, 0, sizeX*sizeY)
	offsetX := 1
	offsetY := 1
	startX := 0
	startY := 0
	loop := min(sizeX, sizeY) / 2

	for ; loop > 0; loop-- {
		x := startX
		y := startY
		for ; x < sizeX-offsetX; x++ {
			ans = append(ans, matrix[y][x])
		}
		for ; y < sizeY-offsetY; y++ {
			ans = append(ans, matrix[y][x])
		}
		for ; x > startX; x-- {
			ans = append(ans, matrix[y][x])
		}
		for ; y > startY; y-- {
			ans = append(ans, matrix[y][x])
		}
		startX++
		startY++
		offsetX++
		offsetY++
	}


	if startX==sizeX-offsetX&&startY==sizeY-offsetY{
		ans = append(ans, matrix[startY][startX])
	}else if startX==sizeX-offsetX {
		for startY<=sizeY-offsetY{
			ans = append(ans, matrix[startY][startX])
			startY++
		}
	}else if startY==sizeY-offsetY {
		for startX<=sizeX-offsetX{
			ans = append(ans, matrix[startY][startX])
			startX++
		}
	}

	return ans
}

File: codes/array/54.spiral-matrix/spiral-matrix.cpp
#include <vector>
#include <algorithm>
using namespace std;
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int startX = 0, startY = 0;
        int offsetX = 1, offsetY = 1;
        int sizeX = matrix[0].size();
        int sizeY = matrix.size();
        int loop = min(matrix.size(), matrix[0].size()) / 2;
        vector<int> ans;
        ans.reserve(sizeX * sizeY);
        while (loop--) {
            int x = startX, y = startY;
            for (; x < sizeX - offsetX; ++x) {
                ans.push_back(matrix[y][x]);
            }
            for (; y < sizeY - offsetY; ++y) {
                ans.push_back(matrix[y][x]);
            }
            for (; x > startX; x--) {
                ans.push_back(matrix[y][x]);
            }
            for (; y > startY; y--) {
                ans.push_back(matrix[y][x]);
            }
            startX++;
            startY++;
            offsetX++;
            offsetY++;
        }
        // 处理因为奇偶性造成的还有一部分没有处理完的情况
        // 剩余中心元素
        if (startX == sizeX - offsetX && startY == sizeY - offsetY) {
            ans.push_back(matrix[startY][startX]);
        }
        // 剩余一列元素
        else if (startX == sizeX - offsetX) {
            while (startY <= sizeY - offsetY) {
                ans.push_back(matrix[startY++][startX]);
            }
        }
        // 剩余一行元素
        else if (startY == sizeY - offsetY) {
            while (startX <= sizeX - offsetX) {
                ans.push_back(matrix[startY][startX++]);
            }
        }
        return ans;
    }
};

int main() {
    Solution s;
    vector<vector<int>> v = {{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, {11, 12, 13, 14, 15, 16, 17, 18, 19, 20}};
    auto ans = s.spiralOrder(v);
}